### PR TITLE
[AMDGPU] Fix sign confusion in performMulLoHiCombine

### PR DESCRIPTION
**Pull request by jayfoad**
SMUL_LOHI and UMUL_LOHI are different operations because the high part
of the result is different, so it is not OK to optimize the signed
version to MUL_U24/MULHI_U24 or the unsigned version to
MUL_I24/MULHI_I24.

### CODE REVIEW

**Comment by jayfoad** [comment]
cc @marekolsak

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp [hunk 1]
```diff
@@ -4360,20 +4361,25 @@ AMDGPUTargetLowering::performMulLoHiCombine(SDNode *N,
...
+  if (Signed) {
+    if (Subtarget->hasMulI24() && isI24(N0, DAG) && isI24(N1, DAG)) {
+      N0 = DAG.getSExtOrTrunc(N0, DL, MVT::i32);
+      N1 = DAG.getSExtOrTrunc(N1, DL, MVT::i32);
+      LoOpcode = AMDGPUISD::MUL_I24;
+      HiOpcode = AMDGPUISD::MULHI_I24;
+    }
   } else {
-    return SDValue();
+    if (Subtarget->hasMulU24() && isU24(N0, DAG) && isU24(N1, DAG)) {
```
Comment (rampitec) [inline llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:4374]: `} else if {`

**Review by rampitec — APPROVED** [review]
LGTM with a nit.

File: llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:4374]: Don't you like symmetry?!

Comment (rampitec) [reply llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:4374]: I do, but I don't like `else { if ...`. But up to you, really.

**Pull request merged**